src/ethereum/frontier/vm/memory.py

```diff
@@ -0,0 +1,112 @@
+"""
+Ethereum Virtual Machine (EVM) Memory
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+EVM memory operations.
+"""
+from ethereum.frontier.vm import Evm
+from ethereum.utils import ceil32
+
+from ...base_types import U256, Bytes
+from .gas import calculate_memory_gas_cost, subtract_gas
+
+
+def extend_memory_and_subtract_gas(
+    evm: Evm, start_position: U256, size: U256
+) -> None:
+    """
+    Extends the size of the memory and
+    substracts the gas amount to extend memory.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    start_position :
+        Starting pointer to the memory.
+    size :
+        Amount of bytes by which the memory needs to be extended.
+    """
+    memory = evm.memory
+    memory_size = U256(len(memory))
+    before_size = ceil32(memory_size)
+    after_size = ceil32(start_position + size)
+    already_paid = calculate_memory_gas_cost(before_size)
+    total_cost = calculate_memory_gas_cost(after_size)
+    if total_cost < already_paid:
+        return None
+    to_be_paid = total_cost - already_paid
+    evm.gas_left = subtract_gas(evm.gas_left, to_be_paid)
+    memory_extend(evm.memory, start_position, size)
+
+
+def memory_extend(memory: bytearray, start_position: U256, size: U256) -> None:
+    """
+    Extends size of memory
+
+    Parameters
+    ----------
+    memory :
+        Memory contents of the EVM.
+    start_position :
+        Starting pointer to the memory.
+    size :
+        Amount by which the memory needs to be extended.
+    """
+    if size == 0:
+        return None
```

Comment:
I think both these functions can be merged into a function called `expand_memory` or `extend_memory`. Since it's already mentioned in the docstring, I don't think it should be part of the function name.

FWIW, the condition `if size == 0: return None` should be present in the `extend_memory_and_subtract_gas` function.

I also prefer using `current_memory_size` and `expanded_memory_size` instead of `before_size` and `after_size` respectively.

Given that I don't think the memory can shrink, the following condition seems unnecessary.
```
if total_cost < already_paid:
    return None
```

Reply:
> Given that I don't think the memory can shrink, the following condition seems unnecessary.

It's true that the memory can't shrink, but you can initially `mload(100)` and later `mload(0)`. For the second `mload`, if I'm not mistaken, that would make `total_cost - already_paid` negative?

Reply:
Ahh, I now understand what's happening here and why those checks were enforced.

@voith I think you need to check if the memory which we are trying to access is already available or not. If it's already available the memory expansion shouldn't happen. Enforcing this check is much more clearer and readable than checking if the new gas is less than the already paid gas.

I leave it up to you whether you want to enforce this check in external code wherever this function is being called or inside this function itself. I weakly prefer having this check outside this function. You can refer to [this condition](https://github.com/ethereum/eth1.0-specs/pull/260/files#diff-5858d8f65876b560db7245474aad6c9acc689d7b6c887a008929cad952b32882R50-R53) in my old PR which actually implements this check.